File: ml-benchmark/synduce/constraints/program/memo_psum.f
Inductive AExpr = int Int | plus {AExpr, AExpr} | minus {AExpr, AExpr};
Inductive NExpr = nint Int | nplus {Int, NExpr, NExpr} | nminus {Int, NExpr, NExpr};

repr = fix (
  \f: NExpr -> AExpr. \e: NExpr.
  match e with
    nint a -> int a
  | nplus {_, a, b} -> plus {f a, f b}
  | nminus {_, a, b} -> minus {f a, f b}
  end
);

memo = \e: NExpr. match e with nint a -> a | nplus {a, _, _} -> a | nminus {a, _, _} -> a end;

is_memo = fix (
  \f: NExpr -> Bool. \e: NExpr.
  match e with
    nint a -> true
  | nplus {n, e1, e2} -> and (== n (+ (memo e1) (memo e2))) (and (f e1) (f e2))
  | nminus {n, e1, e2} -> and (== n (- (memo e1) (memo e2))) (and (f e1) (f e2))
  end
);

spec = fix (
  \f: AExpr -> Int. \e: AExpr.
  match e with
    int a -> a
  | plus {e1, e2} -> + (f e1) (f e2)
  | minus {e1, e2} -> - (f e1) (f e2)
  end
);

target = fix (
  \f: NExpr -> Compress NExpr. \e: NExpr.
  match e with
    nint a -> nint a
  | nplus {s, _, _} -> e /*Avoid recursions of sub-terms*/
  | nminus {s, _, _} -> e
  end
);

/*Customized generator*/
gen = fix (
  \f: AExpr -> NExpr. \e: AExpr.
  match e with
    int a -> nint a
  | plus {a, b} ->
    let res = {f a, f b} in nplus {+ (memo res.1) (memo res.2), res.1, res.2}
  | minus {a, b} ->
    let res = {f a, f b} in nminus {- (memo res.1) (memo res.2), res.1, res.2}
  end
);

main = \e: AExpr. let inp = gen e in
  if is_memo inp then spec (repr (target inp)) else 0;

File: ml-benchmark/synduce/constraints/program/memo_psum.ml
type aexpr =
  | Int of int
  | Plus of aexpr * aexpr
  | Minus of aexpr * aexpr

type nexpr =
  | Nint of int
  | Nplus of int * nexpr * nexpr
  | Nminus of int * nexpr * nexpr

let rec repr e =
  match e with
  | Nint a -> Int a
  | Nplus (_, a, b) -> Plus (repr a, repr b)
  | Nminus (_, a, b) -> Minus (repr a, repr b)

let memo e =
  match e with
  | Nint a -> a
  | Nplus (a, _, _) -> a
  | Nminus (a, _, _) -> a

let rec is_memo e =
  match e with
  | Nint _ -> true
  | Nplus (n, e1, e2) -> (n == memo e1 + memo e2) && (is_memo e1 && is_memo e2)
  | Nminus (n, e1, e2) -> (n == memo e1 - memo e2) && (is_memo e1 && is_memo e2)

let rec spec e =
  match e with
  | Int a -> a
  | Plus (e1, e2) -> spec e1 + spec e2
  | Minus (e1, e2) -> spec e1 - spec e2

val target: nexpr -> nexpr compress
let rec target e =
  match e with
  | Nint a -> Nint a
  | Nplus (_, _, _) -> e
  | Nminus (_, _, _) -> e

let rec gen e =
  match e with
  | Int a -> Nint a
  | Plus (a, b) ->
    let res = (gen a, gen b) in (
      match res with
      | (r1, r2) -> Nplus (memo r1 + memo r2, r1, r2)
    )
  | Minus (a, b) ->
      let res = (gen a, gen b) in
      match res with
      | (r1, r2) -> Nminus (memo r1 - memo r2, r1, r2)

let program e =
  let inp = gen e in
  if is_memo inp then spec (repr (target inp)) else 0